ledger/accountdb.go

```diff
@@ -633,43 +1055,741 @@ func accountsAddNormalizedBalance(tx *sql.Tx, proto config.ConsensusParams) erro
+	}
+}
+
+func (rd *resourcesData) ClearAppParams() {
+	rd.ApprovalProgram = nil
+	rd.ClearStateProgram = nil
+	rd.GlobalState = nil
+	rd.LocalStateSchemaNumUint = 0
+	rd.LocalStateSchemaNumByteSlice = 0
+	rd.GlobalStateSchemaNumUint = 0
+	rd.GlobalStateSchemaNumByteSlice = 0
+	rd.ExtraProgramPages = 0
+	hadHolding := (rd.ResourceFlags & resourceFlagsNotHolding) == resourceFlagsHolding
+	rd.ResourceFlags -= rd.ResourceFlags & resourceFlagsOwnership
+	rd.ResourceFlags &= ^resourceFlagsEmptyApp
+	if rd.IsEmptyAppFields() && hadHolding {
+		rd.ResourceFlags |= resourceFlagsEmptyApp
+	}
+}
+
+func (rd *resourcesData) SetAppParams(ap basics.AppParams, haveHoldings bool) {
+	rd.ApprovalProgram = ap.ApprovalProgram
+	rd.ClearStateProgram = ap.ClearStateProgram
+	rd.GlobalState = ap.GlobalState
+	rd.LocalStateSchemaNumUint = ap.LocalStateSchema.NumUint
+	rd.LocalStateSchemaNumByteSlice = ap.LocalStateSchema.NumByteSlice
+	rd.GlobalStateSchemaNumUint = ap.GlobalStateSchema.NumUint
+	rd.GlobalStateSchemaNumByteSlice = ap.GlobalStateSchema.NumByteSlice
+	rd.ExtraProgramPages = ap.ExtraProgramPages
+	rd.ResourceFlags |= resourceFlagsOwnership
+	if !haveHoldings {
+		rd.ResourceFlags |= resourceFlagsNotHolding
+	}
+	rd.ResourceFlags &= ^resourceFlagsEmptyApp
+	if rd.IsEmptyAppFields() {
+		rd.ResourceFlags |= resourceFlagsEmptyApp
+	}
+}
+
+func (rd *resourcesData) GetAppParams() basics.AppParams {
+	return basics.AppParams{
+		ApprovalProgram:   rd.ApprovalProgram,
+		ClearStateProgram: rd.ClearStateProgram,
+		GlobalState:       rd.GlobalState,
+		StateSchemas: basics.StateSchemas{
+			LocalStateSchema: basics.StateSchema{
+				NumUint:      rd.LocalStateSchemaNumUint,
+				NumByteSlice: rd.LocalStateSchemaNumByteSlice,
+			},
+			GlobalStateSchema: basics.StateSchema{
+				NumUint:      rd.GlobalStateSchemaNumUint,
+				NumByteSlice: rd.GlobalStateSchemaNumByteSlice,
+			},
+		},
+		ExtraProgramPages: rd.ExtraProgramPages,
+	}
+}
+
+func (rd *resourcesData) SetAssetData(ap ledgercore.AssetParamsDelta, ah ledgercore.AssetHoldingDelta) {
+	if ah.Holding != nil {
+		rd.SetAssetHolding(*ah.Holding)
+	} else if ah.Deleted {
+		rd.ClearAssetHolding()
+	}
+	if ap.Params != nil {
+		rd.SetAssetParams(*ap.Params, rd.IsHolding())
+	} else if ap.Deleted {
+		rd.ClearAssetParams()
+	}
+}
+
+func (rd *resourcesData) SetAppData(ap ledgercore.AppParamsDelta, al ledgercore.AppLocalStateDelta) {
+	if al.LocalState != nil {
+		rd.SetAppLocalState(*al.LocalState)
+	} else if al.Deleted {
+		rd.ClearAppLocalState()
+	}
+	if ap.Params != nil {
+		rd.SetAppParams(*ap.Params, rd.IsHolding())
+	} else if ap.Deleted {
+		rd.ClearAppParams()
+	}
+}
+
+func accountDataResources(
+	ctx context.Context,
+	accountData *basics.AccountData, rowid int64,
+	outputResourceCb func(ctx context.Context, rowid int64, cidx basics.CreatableIndex, rd *resourcesData) error,
+) error {
+	// does this account have any assets ?
+	if len(accountData.Assets) > 0 || len(accountData.AssetParams) > 0 {
```

Comment:
Why check these lengths, since you're about to range over them, skipping them if len == 0 anyway?

(Same question below for Apps)

Reply:
ok, removed.